pinot/data/datasets.py

```diff
@@ -207,3 +423,196 @@ def filter_by_time(self, after="1989-06-04", before="2666-12-31"):
+
+            # specification for multiple
+            multiple_concentration_col_name='f_concentration_uM',
+            multiple_measurement_col_name='f_inhibition_list',
+
+            # specification for signel
+            single_concentrations=[20, 50],
+            single_col_names=[
+                'f_inhibition_at_20_uM',
+                'f_inhibition_at_50_uM',
+            ],
+
+        ):
+
+        def _from_csv():
+            # read single and multiple data
+            df_single = pd.read_csv(single_path)
+            df_multiple = pd.read_csv(multiple_path)
+
+            # merge it
+            df = df_single.merge(
+                right=df_multiple,
+                how='left',
+                on='SMILES'
+            )
+
+            # filter it
+            df = df.filter(items=[
+                smiles_col_name,
+                multiple_concentration_col_name,
+                multiple_measurement_col_name,
+                *single_col_names,
+                ])
+
+            df['cs_single'] = np.nan
+            df['ys_single'] = np.nan
+
+            df = df.rename(
+                columns={
+                    multiple_concentration_col_name: 'cs_multiple',
+                    multiple_measurement_col_name: 'ys_multiple'
+                }
+            )
+
+            def flatten_multiple(record):
+                cs = record['cs_multiple']
+                ys = record['ys_multiple']
+
+                if isinstance(cs, str):
+                    cs = eval(cs)
+                    ys = eval(ys)
+
+                    cs = [x for c in cs for x in c]
+                    ys = [x for y in ys for x in y]
+
+                    record['cs_multiple'] = cs
+                    record['ys_multiple'] = ys
+
+                return record
+
+            df.apply(flatten_multiple, axis=1)
+
+            def flatten_single(record):
+                cs = single_concentrations
+                ys = [record[name] for name in single_col_names]
+
+                record['cs_single'] = cs
+                record['ys_single'] = ys
+
+                return record
+
+            df = df.apply(flatten_single, axis=1)
+
+            self.ds = df.to_dict(orient='records')
+
+            for record in self.ds:
+                record['g'] = pinot.graph.from_rdkit_mol(
+                    Chem.MolFromSmiles(record['SMILES'])
+                )
+
+            # print(self.ds)
+
+            return self
+
+        return _from_csv
+
+    @staticmethod
+    def all_available_pairs(xs):
+        # initialize return lists
+        gs = []
+        cs = []
+        ys = []
+
+        for x in xs: # loop through the data
+            # get the graph
+            g = x['g']
+
+            # get the single point measurements
+            ys_single = x['ys_single']
+            cs_single = x['cs_single']
+
+            # get the multiple point measurements
+            ys_multiple = x['ys_multiple']
+            cs_multiple = x['cs_multiple']
```

Comment:
I remain confused by the distinction we're making between "single" and "multiple" point measurements. Why can't 20 or 50 uM just be yet another concentration/in `c`?

Reply:
Ah, ok so you're appending both multiple and single to `cs`.

Reply:
yup. this could be made more generalize by feeding a dictionary of measurement types, rather than `single` and `multiple`